tests/api2/test_audit_nfs.py

```diff
@@ -0,0 +1,153 @@
+import os
+import sys
+
+import pytest
+from middlewared.service_exception import CallError
+from middlewared.test.integration.assets.pool import dataset
+from middlewared.test.integration.utils import call
+from middlewared.test.integration.utils.audit import expect_audit_method_calls
+
+sys.path.append(os.getcwd())
+from functions import DELETE, POST, PUT
+
+REDACTED_SECRET = '********'
+
+
+@pytest.fixture(scope='module')
+def nfs_audit_dataset(request):
+    with dataset('audit-test-nfs') as ds:
+        try:
+            yield ds
+        finally:
+            pass
+
+
+@pytest.mark.parametrize('api', ['ws', 'rest'])
+def test_nfs_config_audit(api):
+    '''
+    Test the auditing of NFS configuration changes
+    '''
+    bogus_user = 'bogus_user'
+    bogus_password = 'boguspassword123'
+    initial_nfs_config = call('nfs.config')
+    try:
+        # CREATE
+        with expect_audit_method_calls([{
+            'method': 'nfs.add_principal',
+            'params': [
+                {
+                    'username': bogus_user,
+                    'password': REDACTED_SECRET,
+                }
+            ],
+            'description': f'Add NFS principal {bogus_user}',
+        }]):
+            payload = {
+                'username': bogus_user,
+                'password': bogus_password,
+            }
+            # The 'add' will fail, but the audit check should pass
+            if api == 'ws':
+                with pytest.raises(CallError):
+                    call('nfs.add_principal', payload)
+            elif api == 'rest':
+                result = POST('/nfs/add_principal/', payload)
+                assert result.status_code != 200, result.text
+            else:
+                raise ValueError(api)
+        # UPDATE
+        payload = {
+            'mountd_log': not initial_nfs_config['mountd_log'],
+            'mountd_port': 618,
+            'protocols': ["NFSV4"]
+        }
+        with expect_audit_method_calls([{
+            'method': 'nfs.update',
+            'params': [payload],
+            'description': 'Update NFS configuration',
+        }]):
+            if api == 'ws':
+                call('nfs.update', payload)
+            elif api == 'rest':
+                result = PUT('/nfs/', payload)
+                assert result.status_code == 200, result.text
+            else:
+                raise ValueError(api)
+    finally:
+        # Restore initial state
+        restore_payload = {
+            'mountd_log': initial_nfs_config['mountd_log'],
+            'mountd_port': initial_nfs_config['mountd_port'],
+            'protocols': initial_nfs_config['protocols']
+        }
```

Comment:
Does this undo the `nfs.add_principal` ?  If so, maybe add a comment to that effect to clarify.

Reply:
I added a comment that the `add_principal` would call would fail.  So , no setting to undo there.

Reply:
OK